src/dmd/lexer.d

```diff
@@ -2284,6 +2284,8 @@ class Lexer
     final Loc loc() pure @nogc
     {
         scanloc.charnum = cast(uint)(1 + p - line);
+        version(ExtraLocation)
+            scanloc.offset = cast(uint)(line - base) + scanloc.charnum - 1;
```

Comment:
I don't think you have to do this complicated calculation. The offset is more or less already stored in the lexer. You can have a look at how it's stored for inside a token.